pkg/coscheduling/coscheduling.go

```diff
@@ -0,0 +1,267 @@
+	podGroupName, minAvailable, namespace := cs.checkPod(p)
+	if podGroupName == "" || minAvailable <= 1 {
+		return framework.NewStatus(framework.Success, ""), 0
+	}
+
+	running := cs.calculateRunningPods(podGroupName, namespace)
+	waiting := cs.calculateWaitingPods(podGroupName, namespace)
+
+	if running+waiting+1 < minAvailable {
+		klog.V(3).Infof("The count of podGroup %v is not up to minAvailable(%d) in Permit: running(%d), waiting(%d)",
+			namespace+"/"+podGroupName, minAvailable, running, waiting)
+		return framework.NewStatus(framework.Wait, ""), time.Duration(minAvailable-running-waiting-1/minAvailable) * PermitWaitingTime
+	}
+
+	klog.V(3).Infof("The count of podGroup %v is up to minAvailable(%d) in Permit: running(%d), waiting(%d)",
+		podGroupName, minAvailable, running, waiting)
+	cs.FrameworkHandle.IterateOverWaitingPods(func(waitingPod framework.WaitingPod) {
+		if waitingPod.GetPod().Namespace == namespace && waitingPod.GetPod().Labels[PodGroupName] == podGroupName {
+			klog.V(3).Infof("Permit allows the pod: %v/%v", podGroupName, waitingPod.GetPod().Name)
+			waitingPod.Allow(cs.Name())
+		}
+	})
+
+	delete(cs.PodGroupInfos, namespace+"/"+podGroupName)
+	return framework.NewStatus(framework.Success, ""), 0
+}
+
+// Unreserve is called by the scheduling framework when a reserved pod was
+// rejected in a later phase.
+func (cs *Coscheduling) Unreserve(ctx context.Context, state *framework.CycleState, p *v1.Pod, nodeName string) {
+	podGroupName, exist := p.Labels[PodGroupName]
+	if !exist {
+		return
+	}
+
+	cs.FrameworkHandle.IterateOverWaitingPods(func(waitingPod framework.WaitingPod) {
+		if waitingPod.GetPod().Namespace == p.Namespace && waitingPod.GetPod().Labels[PodGroupName] == podGroupName {
+			klog.V(3).Infof("Unreserve rejects the pod: %v/%v", podGroupName, waitingPod.GetPod().Name)
+			waitingPod.Reject(cs.Name())
+		}
+	})
+}
+
+func (cs *Coscheduling) checkPod(p *v1.Pod) (string, int, string) {
+	podGroupName, exist := p.Labels[PodGroupName]
+	if !exist {
+		return "", 0, p.Namespace
+	}
+	minAvailable, exist := p.Labels[PodGroupMinAvailable]
+	if !exist {
+		return "", 0, p.Namespace
+	}
+	minNum, err := strconv.Atoi(minAvailable)
+	if err != nil {
+		klog.Error(err)
+		return "", 0, p.Namespace
+	}
+	klog.V(3).Infof("Pod is in a podGroup: %v/%v, minAvailable is %v", podGroupName, p.Name, minNum)
+	return podGroupName, minNum, p.Namespace
+}
+
+func (cs *Coscheduling) calculateTotalPods(podGroupName, namespace string) int {
+	// Returns the list of pods by the podGroupName.
+	selector := labels.Set(labels.Set{PodGroupName: podGroupName}).AsSelector()
+	pods, err := cs.PodLister.Pods(namespace).List(selector)
+	if err != nil {
+		klog.Error(err)
+		return 0
+	}
+	return len(pods)
+}
+
+func (cs *Coscheduling) calculateAllocatedPods(podGroupName, namespace string) int {
+	allocated := 0
```

Comment:
nit: you can just use the `Pods` returned by `FilteredList()`, and then return `len(pods)`. (applies elsewhere)